freegs/reconstruction.py

```diff
@@ -0,0 +1,803 @@
+        if indexing == 'coils':
+            if self.use_VesselCurrents:
+                return Vector[-(self.tokamak.n_coils+self.tokamak.n_basis):-self.tokamak.n_basis]
+            else:
+                return Vector[-self.tokamak.n_coils:]
+
+        if indexing == 'vessel':
+            return Vector[-self.tokamak.n_basis:]
+
+    # Reconstruction Algorithm
+    def reconstruct(self):
+        """
+        Parameters
+        ----------
+        tokamak - tokamak object
+        M - measurements
+        sigma - measurement uncertainties
+        pprime_order - number of polynomial coefficients for pprime model
+        ffprime_order - number of polynomial coefficients for ffprime model
+        tolerance - value beneath which the chi squared value will stop the iteration
+        Gplasma - Plasma Response Greens matrix
+        Gcoil - Coil Greens matrix
+        Gvessel - Eigenbasis response Greens matrix
+        show - option to display psi through the iterations
+        pause - time to wait between iterations
+        use_VesselCurrents - Used when tokamak has vessel filaments. Allows calculation of induced currents in tokamak wall
+        use_VerticalControl - option for use of the vertical control found in ferron's paper
+        use_CurrentInitialisation - option for using finite element method to optimise initial guess for jtor
+        """
+
+        # start iterative loop
+        self.it = 0  # iteration number
+        convergence = False
+        while not convergence:
+
+            # Initialisation
+            if self.it == 0:
+                jtor_2d = None
+
+                if self.show:
+                    import matplotlib.pyplot as plt
+                    from .plotting import plotEquilibrium
+
+                    if self.pause > 0.0:
+                        # No axis specified, so create a new figure
+                        fig, axs = plt.subplots(1,2, sharex=True, sharey=True)
+
+                # Performs plasma current density initialisation
+                if self.use_CurrentInitialisation:
+                    jtor_1d = self.initialise_plasma_current()
+                    self.Jtor = np.reshape(jtor_1d, (self.nx, self.ny),order='F')
+
+                F = self.get_fittingWeightVector()
+
+            else:
+
+                if self.show:
+                    self.plot_equilibria(fig, axs)
+
+                # Use B & C to calculate Jtor matrix
+                jtor_1d = Basis @ self.get_index(self.c, 'plasma')
+                self.Jtor = np.reshape(jtor_1d, (self.nx, self.ny),order='F')
+
+            # Calculate Psi values with elliptical solver
+            x_2d, mask = self.get_psi_norm()
+            self.psi_norm = x_2d.flatten('F')
+
+            # Calculate B and apply mask
+            Basis = self.get_BasisMatrix()
+            Basis = self.apply_mask(Basis, mask)
+
+            # Calculating operator matrix A
+            A = self.Gplasma @ Basis
```

Comment:
what is A, better name plz

Reply:
how are you ensuring Gplasma is updated?

Reply:
i will make it to Aplasma, to match the mathematical document. Why would Gplasma need updating?